### PR TITLE
Updated types for Lolex to the current version (3)

### PR DESCRIPTION
**Pull request by zyishai**
Please fill in this template.

- [X] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [X] Add or edit tests to reflect the change. (Run with `npm test`.)
- [X] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [X] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [X] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

If changing an existing definition:
- [X] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/sinonjs/lolex and https://github.com/sinonjs/lolex/issues/220.

### CODE REVIEW

**Comment by benjamingr** [comment]
As lolex core - this looks good.

File: types/lolex/index.d.ts
Comment (JoshuaKGoldberg) [inline types/lolex/index.d.ts]: Nit: if the generic is no longer here, you can just remove the `type` line.

File: types/lolex/index.d.ts
Comment (JoshuaKGoldberg) [inline types/lolex/index.d.ts]: Instead of removing the `<TClock extends Clock>`, could you change it to `<TClock extends InstalledClock = InstalledClock>`? It used to be useful to specify which type of clock it was (browser vs node).

File: types/lolex/lolex-tests.ts
Comment (JoshuaKGoldberg) [inline types/lolex/lolex-tests.ts]: Are the browser and node clocks no longer able to to `uninstall();`?

**Review by JoshuaKGoldberg — CHANGES_REQUESTED** [review]
Reviewing from just a TypeScript perspective since benjamingr approved the content changes - mostly great, with a few touchup requests.

File: types/lolex/lolex-tests.ts
Comment (benjamingr) [reply types/lolex/lolex-tests.ts]: browser clocks are able to uninstall.

Comment (zyishai) [reply types/lolex/lolex-tests.ts]: @JoshuaKGoldberg in Lolex source, the method `uninstall()` is only added to clocks that created with `install()`, but clocks that were created with `createClock()` have no `uninstall()`. That's why I removed the uninstallation of `browserClock` and `nodeClock` and added `installedClock` (line 31) and called `uninstall()` on it (line 124).

File: types/lolex/index.d.ts
Comment (zyishai) [reply types/lolex/index.d.ts]: Sure, will do. But could you explain to me what the = does? I didn't saw syntax like that in the documentation.

Comment (JoshuaKGoldberg) [reply types/lolex/index.d.ts]: These are [Generic parameter defaults](https://github.com/Microsoft/TypeScript/wiki/What%27s-new-in-TypeScript#generic-parameter-defaults). If the user doesn't provide an explicitly value for the generic, it'll default to whatever that value is. The docs are a little out of date on these newer things 🙁 but it's super nifty!

Come to think of it, maybe the `TClock` generic type should extend `Clock`, and the return type be `TClock & InstalledClock`? Something like _(but with better names)_:

```typescript
type InstalledClockBase = {
    uninstall: () => void;
    methods: FakeMethod[];
};

function install<TClock extends Clock>(opts?: LolexInstallOpts): TClock & InstalledClockBase;

type InstalledClock = InstalledClockBase & Clock;
```

```typescript
const browserClock = install<BrowserClock>(); // BrowserClock & InstalledClock;
const clock = install(); // InstalledClock
```

**Comment by thejohnfreeman** [comment]
This repository cannot be used directly as a source for the `@types/lolex` package. I tried checking out @zyishai's branch to see if I could publish it, but the release process is far too complicated if you don't want to publish every single package under the scope `@types`. Can we get instructions on how to publish just this package? Or can we get this PR merged? @JoshuaKGoldberg ? @benjamingr ? Pretty please? Let me know if you need some help and how I can provide it.

**Comment by thejohnfreeman** [comment]
I figured out how to [cobble it together by hand](https://github.com/thejohnfreeman/types-lolex) and will use that while we wait for this PR to be merged.

```javascript
// package.json
"@types/lolex": "https://github.com/thejohnfreeman/types-lolex.git#master"
```

**Comment by JoshuaKGoldberg** [comment]
@thejohnfreeman this repository will release the new types automatically once this PR is merged. That'll happen automatically once >=1 person has signed off and nobody is requesting changes.

...but taking another glance, it looks like I dropped the ball and should have re-reviewed and approved! 😲 sorry for not coming back to this!

/cc @zyishai

**Review by JoshuaKGoldberg — APPROVED** [review]

**Pull request merged**
